src/findall_window.jl

```diff
@@ -0,0 +1,51 @@
+"""
+    findall_window(f, img::AbstractArray{T,N}, window::Dims{N}; allow_edges::(true...,))
+
+Return the list of window-centers for which `f(V, basepoint)` returns `true` for windowed-views `V`
+"centered" on `basepoint`. At the edge of `img`, `basepoint` may not actually be in the center of `V`.
+"""
+function findall_window(f::F, img::AbstractArray{T,N}, window::Dims{N}; allow_edges::NTuple{N,Bool}=ntuple(_->true, N)) where {F,T<:Union{Gray,Number},N}
+    basepoints = Vector{CartesianIndex{N}}(undef, 0)
+    Iedge = CartesianIndex(map(!, allow_edges))
+    R0 = CartesianIndices(img)
+    R = clippedinds(R0, Iedge)
+    halfwindow = CartesianIndex(map(x -> x >> 1, window))
+    for i in R
+        Rview = _colon(i-halfwindow, i+halfwindow) ∩ R0
+        if f(OffsetArray(@inbounds(view(img, Rview)), Rview.indices), i)
+            push!(basepoints, i)
+        end
+    end
+    return basepoints
+end
+
+"""
+    all_window(f, V, basepoint, excludepoint=nothing)
+
+Returns `true` if `f(img[centerpoint], img[otherpoint])` for all indices `otherpoint`.
+Optionally exclude a single point (e.g., `basepoint` itself) by setting `excludepoint`.
+"""
+@inline function all_window(f::F, V::AbstractArray, basepoint::CartesianIndex, excludepoint::Union{Nothing,CartesianIndex}=nothing) where F
+    @inbounds ref = V[basepoint]
+    @inbounds for (i, v) in pairs(V)
+        i == excludepoint || f(ref, v) || return false
```

Comment:
I'm expecting `f(ref, v)` is scalar operation and is type stable, but if I change to

```julia
i == excludepoint || ref == v || return false
```

I get 

```julia
julia> @btime ImageFiltering.findmax_window($A, $(ImageFiltering.default_window(A)));
  74.347 μs (1 allocation: 80 bytes)
```

Surprisingly less allocations, do you have any ideas?

Reply:
Presumably the output array is empty? So it's probably just `push!` reallocating the output buffer.

Reply:
With

```julia
function findall_window(f::F, img::AbstractArray{T,N}, window::Dims{N}; allow_edges::NTuple{N,Bool}=ntuple(_->true, N)) where {F,T<:Union{Gray,Number},N}
    buffer_size = min(512, length(img)) # a heuristic threshold that to work around the overhead of `push!`
    basepoints = Vector{CartesianIndex{N}}(undef, buffer_size)
    Iedge = CartesianIndex(map(!, allow_edges))
    R0 = CartesianIndices(img)
    R = clippedinds(R0, Iedge)
    halfwindow = CartesianIndex(map(x -> x >> 1, window))
    i_positive = 0
    @inbounds for i in R
        Rview = _colon(i-halfwindow, i+halfwindow) ∩ R0
        if f(OffsetArray(view(img, Rview), Rview.indices), i)
            i_positive += 1
            if length(basepoints) < i_positive
                resize!(basepoints, min(2length(basepoints), length(img)))
            end
            basepoints[i_positive] = i
        end
    end
    return resize!(basepoints, i_positive)
end
```

I'm getting closer:

```julia
julia> @btime ImageFiltering.findmax_window($A, $(ImageFiltering.default_window(A)));
  146.643 μs (4 allocations: 60.31 KiB) # after
  167.950 μs (11 allocations: 64.47 KiB) # before
```

Reply:
`sizehint!` does the same thing and is much simpler.

Reply:
And these optimizations could be applied to `findlocalextrema`, widening the gap once again. (The same comment applies to my own `i == excludepoint || ...` change.)

I think the key question is any overhead from creating the wrapper, which `findlocalextrema` doesn't do.